### PR TITLE
Move addcmul to Aten(CUDA)

### PR DESCRIPTION
**Pull request by ifedan**
https://github.com/pytorch/pytorch/issues/22797

### CODE REVIEW

**Comment by ifedan** [comment]
Performance:

**pytorch:master**
```
sizes = (5, 5) mean = 0.011118258660038313 std = 0.00038343089200118156
sizes = (5, 100) mean = 0.011211889407907924 std = 0.0002408705379617397
sizes = (5, 10000) mean = 0.011075803001100818 std = 0.00010591437464095989
sizes = (100, 5) mean = 0.01122387358918786 std = 0.0001789595290926239
sizes = (100, 100) mean = 0.011082999020194014 std = 0.0002470536335692196
sizes = (100, 10000) mean = 0.04222017868111531 std = 8.095860378991753e-05
sizes = (10000, 5) mean = 0.010851199583460888 std = 0.00036147434389757706
sizes = (10000, 100) mean = 0.04180279102486869 std = 0.00010877182560183754
sizes = (10000, 10000) mean = 3.234245077241212 std = 0.000514372900331391
```
**ifedan:addcmul**
```
sizes = (5, 5) mean = 0.007850277315204343 std = 0.00011302040360778973
sizes = (5, 100) mean = 0.008031459370007118 std = 8.580753278991458e-05
sizes = (5, 10000) mean = 0.007864954726149639 std = 4.363541661316275e-05
sizes = (100, 5) mean = 0.008011385022352139 std = 6.75816184693509e-05
sizes = (100, 100) mean = 0.008333299619456133 std = 0.0005226180366067936
sizes = (100, 10000) mean = 0.02832021564245224 std = 8.559916979319577e-05
sizes = (10000, 5) mean = 0.007843895892923078 std = 8.487525835236113e-05
sizes = (10000, 100) mean = 0.028705650319655735 std = 0.0006553430335311091
sizes = (10000, 10000) mean = 2.214352016026775 std = 0.043058109890723915
```

File: aten/src/THC/generic/THCTensorMathPointwise.cu [hunk 1]
```diff
@@ -609,30 +609,6 @@ void THCTensor_(cfmod)(THCState *state, THCTensor *self, THCTensor *src1, THCTen
...
-  else
-  {
-    THArgCheck(THCTensor_(nElement)(state, self_) == THCTensor_(nElement)(state, src1),
-               1, "sizes do not match");
-  }
-
-  THArgCheck(THCTensor_(nElement)(state, src1) == THCTensor_(nElement)(state, src2),
-             3, "sizes do not match");
-
-  if (!THC_pointwiseApply3<scalar_t, scalar_t, scalar_t>(state, self_, src1, src2, TensorAddCMulOp<scalar_t>(value))) {
```
Comment (VitalyFedyunin) [inline aten/src/THC/generic/THCTensorMathPointwise.cu]: Please also remove `TensorAddCMulOp` from the code base.

File: test/test_torch.py [hunk 2]
```diff
@@ -4016,6 +4016,17 @@ def _spawn_method(self, method, arg):
         with mp.Pool(1) as pool:
             self.assertTrue(pool.map(method, [arg]))
 
+    def test_addcmul(self):
+        for device in torch.testing.get_all_device_types():
+            for dtype in [torch.float, torch.double]:
```
Comment (VitalyFedyunin) [inline test/test_torch.py]: Test integral types as well.

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Overall good, just tiny things left to be done.

**Comment by syed-ahmed** [comment]
I would also suggest posting some bandwidth numbers. On V100, you should peak at about 750-800 GB/s. Here is a script to calculate bandwidth numbers: https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#cuda-development-tips

**Comment by ifedan** [comment]
> I would also suggest posting some bandwidth numbers. On V100, you should peak at about 750-800 GB/s. Here is a script to calculate bandwidth numbers: https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#cuda-development-tips
```
uniform, size, elements 65536 forward 1.2671947479248047e-05 bandwidth (GB/s) 20.68695442664158
uniform, size, elements 131072 forward 1.2950897216796875e-05 bandwidth (GB/s) 40.48275507275405
uniform, size, elements 262144 forward 1.783132553100586e-05 bandwidth (GB/s) 58.80527491782324
uniform, size, elements 524288 forward 2.301931381225586e-05 bandwidth (GB/s) 91.1040188732056
uniform, size, elements 1048576 forward 3.739118576049805e-05 bandwidth (GB/s) 112.17360227262641
uniform, size, elements 2097152 forward 6.335020065307617e-05 bandwidth (GB/s) 132.41643930914157
uniform, size, elements 4194304 forward 0.00012045621871948243 bandwidth (GB/s) 139.2806131418641
uniform, size, elements 8388608 forward 0.00021587133407592772 bandwidth (GB/s) 155.4371827256972
uniform, size, elements 16777216 forward 0.00041317224502563477 bandwidth (GB/s) 162.4234560959832
uniform, size, elements 33554432 forward 0.0007810091972351074 bandwidth (GB/s) 171.85166125463232
```

File: test/test_torch.py (hunk 2 above)
Comment (ifedan) [reply test/test_torch.py]: Added

File: aten/src/THC/generic/THCTensorMathPointwise.cu (hunk 1 above)
Comment (ifedan) [reply aten/src/THC/generic/THCTensorMathPointwise.cu]: Removed

**Comment by syed-ahmed** [comment]
@ifedan Assuming those numbers are on V100, I'd look into a bit why the bandwidth is low. Last time I benchmarked BinaryOps and Distribution ops (which uses grid-stride loop) they were all close to peak bandwidth (~800 GB/s). I guess this is not required for this PR and can be opened as a separate issue as long as the bandwidth numbers before and after are the same.

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**